scripts/abi_to_rc/abi_to_rc.py

```diff
@@ -0,0 +1,155 @@
+    if not os.path.isfile(args.abi_file):
+        abi_filename = os.path.split(args.abi_file)[1]
+        print("ERROR: %s could not be found" % abi_filename)
+        exit(1)
+    if not os.path.isfile(os.path.join(os.path.dirname(sys.argv[0]), _RC_OVERVIEW)):
+        print("ERROR: %s could not be found" % _RC_OVERVIEW)
+        exit(1)
+    if not os.path.isfile(os.path.join(os.path.dirname(sys.argv[0]), _RC_ACTION)):
+        print("ERROR: %s could not be found" % _RC_ACTION)
+        exit(1)
+
+# gets actions, inputs and input types from abi file
+def get_actions_inputs_types():
+    abi_file = open(args.abi_file,'r')
+    abi_text = abi_file.read()
+    abi_file.close()
+    abi_json = json.loads(abi_text)
+    actions_json = abi_json['actions']
+    for obj in actions_json:
+        actions.append(obj)
+    structs_json = abi_json['structs']
+    for action in actions:
+        inputs[action['name']] = []
+        types[action['name']] = []
+    for struct in structs_json:
+        for action in actions:
+            if struct['name'] == action['type']:
+                for field in struct['fields']:
+                    inputs[action['name']].append(field['name'])
+                    types[action['name']].append(field['type'])
+
+# builds rows for the table
+def build_table_rows():
+    table_rows = []
+    for action in actions:
+        action_string = "`" + action['name'] + "`"
+        input_string = ""
+        input_list = []
+        type_string = ""
+        type_list =[]
+        if len(inputs[action['name']]) >= 1:
+            for name in inputs[action['name']]:
+                input_list.append("`" + name + "`")
+            input_string = '<br/>'.join(input_list)
+        else:
+            input_string = "`" + action['type'] + "`"
+        if len(types[action['name']]) >= 1:
+            for name in types[action['name']]:
+                type_list.append("`" + name + "`")
+            type_string = '<br/>'.join(type_list)
+        else:
+            type_string = "`" + action['type'] + "`"
+        table_rows.append('| ' + action_string + ' | ' + input_string + ' | ' + type_string + ' |')
+    return table_rows
+
+# generates an overview ricardian contract from the overview template
+def generate_rc_overview_file():
+    tr = build_table_rows()
+    plural = {  'action': 'actions' if len(actions) > 1 else 'action',
+                'input': 'inputs' if len(inputs) > 1 else 'input',
+                'type': 'types' if len(types) > 1 else 'type'      }
+    abi_file_name = os.path.split(args.abi_file)[1]
+    contract_name = os.path.splitext(abi_file_name)[0]
+    rc_file_name = contract_name + '-rc.md'
+    dirname = os.path.split(args.abi_file)[0]
+    rc_file = open(os.path.join(dirname, rc_file_name),"w+")
+    with open(os.path.join(os.path.dirname(sys.argv[0]), _RC_OVERVIEW)) as fp:
+        overview_template = Template(fp.read())
+        rc_file.write(overview_template.substitute( contract=contract_name,
```

Comment:
Rather than use the **kwds form of `substitute()`, it would be considerably less verbose to simply pass a dictionary.  You nearly have the one you need already.  Rename `plural` to something like `subs` and insert the other four things into it. `'contract': contract_name` can just be added to the initial declaration.  Then add the titlecase versions with a one-liner: `subs.update([(k+'_header',v.title()) for k,v in subs.copy().items()])`.  You'll end up with an extra `contract_header` entry, but Template will politely ignore it for you.  Then you can just call `overview_template.substitute(subs)`.

Admittedly, reading Python list comprehensions takes some practice.  Upside down `for` loops aren't exactly what you expect, but you get used to it.